### ISSUE TITLE
bip32 public key derivations (starting with capital M) aren't handled correctly

### ISSUE THREAD

**Issue by tphyahoo** [jprupp/haskoin-core#196]
From Network.Haskoin.Crypto.ExtendedKeys.Units.derivePrvPathVectors

```
, ( xprv, "m/8'", hardSubKey xprv $ XKeyHardIndex 8 )
, ( xprv, "M/8'", hardSubKey xprv $ XKeyHardIndex 8 ) 
```

this test passes if behavior is identical for M as for m.

I believe the convention is for M to derive a public key. It's a bit sketchy in the bip32 spec but

https://github.com/GemHQ/money-tree has this, and the current repo also has in 

Network.Haskoin.Crypto.ExtendedKeys.parrsePath: 

```
pEnd = case x of
    ""  -> Just Deriv
    "m" -> Just DerivPrv
    "M" -> Just DerivPub
    _   -> Nothing
```

 So I think the above test needs fixing.

Right?

**Comment by tphyahoo** [comment]
My solution for the above issue was to replace derivePrvPathVectors with deriveHardPrvPathVectors.

See 

https://github.com/tphyahoo/haskoin/commit/b6574dc5961c820bf3147adc4325bb514e2e4e54#diff-b4a7936eaf19e08d409c64c9fb5a6a4a

and ctrl-f for these functions in the large diff.

Test strings must now parse as hard path. Soft paths, or DerivPath parsable strings are no longer allowed.

I think library users will probably usually want to use deriveBip32Path (which replaces derivePathE, which I left as a synonym for now but would like to deprecate), which is the most flexible method. For deriveBip32Path, you can start with either a public or private key. You use new data type Bip32XKey which allows either pub or priv as start point. And since DerivPath transformations can point to either private key (with lower case m) or public key (with upper case M), the result type is again Bip32XKey. Well actually, the result type is Maybe Bip32XKey to allow failures in the case of impossible derivations. 

Impossible derivations include: m/1/2/3 starting with pub key.  Or M/1'/2'/3' (a hard path) starting with pub key.

(Perhaps this should be added to haddock.)

For new behavior of derivePathE, ctrl-f derivePathVectors in above linked commit.

**Comment by tphyahoo** [comment]
This can be closed as soon as #200 is accepted to haskoin-core.

**Comment by tphyahoo** [comment]
In light of 

https://github.com/haskoin/haskoin/pull/243 and https://github.com/haskoin/haskoin/pull/242

I'm satisfied that M/ paths are handled correctly for applyPath.

I don't think there are other functions that take M/ paths. If there are, perhaps these can be double checked for haddock+tests as well.

UPDATE: applyPath is the only function that takes ParsedPath and a key, so this looks good to close.

I would still like to have a type for ExtKey paths that captures hardness or softness at the type level, and I have some branches where I'm messing with this, but new issues can be opened if I come up with something I like, and I will also take care to coexist peacefully with existing type machinery.
